compiler/src/iree/compiler/Dialect/Flow/Transforms/Canonicalizer.cpp

```diff
@@ -17,6 +18,82 @@ namespace mlir::iree_compiler::IREE::Flow {
 
 namespace {
 
+/// Folds a chain of `tensor.pad` ops with the same constant padding value.
+///
+/// Example:
+///
+/// ```mlir
+///   %0 = tensor.pad %0 low[0, 1] high[0, 2] {
+///       tensor.yield %val
+///     } : tensor<1x2xf32> to tensor<2x5xf32>
+///   %res = tensor.pad %0 low[0, 2] high[3, 0] {
+///       tensor.yield %val
+///     } : tensor<1x5xf32> to tensor<5x7xf32>
+/// ```
+///
+/// folds into:
+///
+/// ```mlir
+///   %res = tensor.pad %-1 low[0, 3] high[3, 2] {
+///       tensor.yield %val
+///     } : tensor<1x2xf32> to tensor<5x7xf32>
+/// ```
+///
+/// NOTE: This wasn't sent upstream as a canonicalization due to the use of
+/// the Affine dialect.
+struct FoldConsecutiveConstantPadding : public OpRewritePattern<tensor::PadOp> {
+  using OpRewritePattern<tensor::PadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::PadOp padOp,
+                                PatternRewriter &rewriter) const override {
+    if (padOp.getNofold()) {
+      return failure();
+    }
+    auto producerPad = padOp.getSource().getDefiningOp<tensor::PadOp>();
+    if (!producerPad || producerPad.getNofold())
+      return failure();
```

Comment:
nit: wrap with braces. I think we can notify useful message here, it's quite useful in `-debug`.

```suggestion
    if (!producerPad || producerPad.getNofold()) {
      return rewriter.notifyMatchFailure(padOp, "producer is not a foldable tensor.pad op");
    }
```